rustls/src/server/verifier_builder.rs

```diff
@@ -0,0 +1,427 @@
+    /// Build a client certificate verifier that allow all clients that present a client
+    /// certificate that is issued by a trusted CA, or anonymous clients that do not offer a
+    /// client certificate at all. If configured, revocation checking with CRLs
+    /// will be done when a client presents a certificate.
+    pub fn build(self) -> Arc<dyn ClientCertVerifier> {
+        Arc::new(AllowAnyAnonymousOrAuthenticatedClient {
+            inner: AllowAnyAuthenticatedClient::new(self.state.roots, self.state.crls),
+        })
+    }
+}
+
+impl fmt::Debug for ClientAuthenticationOptional {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("ClientAuthenticationOptional")
+            .field("roots", &self.roots)
+            // TODO(XXX): replace when OwnedCertRevocationList implements Debug.
+            .field("crls", if self.crls.is_empty() { &"no" } else { &"yes" })
+            .finish()
+    }
+}
+
+/// A client certificate verifier builder state where the server will perform no client
+/// authentication.
+#[derive(Clone, Debug)]
+pub struct NoClientAuthentication(pub(crate) ());
+
+impl ClientCertVerifierBuilder<NoClientAuthentication> {
+    /// Build a "no operation" client certificate verifier that will allow all clients to connect,
+    /// performing no client certificate authentication or revocation checking.
+    pub fn build(self) -> Arc<dyn ClientCertVerifier> {
+        Arc::new(NoClientAuth {})
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        server::{UnparsedCertRevocationList, WebpkiClientVerifier},
+        CertRevocationListError, Certificate, RootCertStore,
+    };
+
+    fn test_crls() -> Vec<UnparsedCertRevocationList> {
+        [
+            include_bytes!("../../../test-ca/ecdsa/client.revoked.crl.pem").as_slice(),
+            include_bytes!("../../../test-ca/rsa/client.revoked.crl.pem").as_slice(),
+        ]
+        .iter()
+        .map(|pem_bytes| {
+            UnparsedCertRevocationList(
+                rustls_pemfile::crls(&mut &pem_bytes[..])
+                    .unwrap()
+                    .first()
+                    .unwrap()
+                    .to_vec(),
+            )
+        })
+        .collect()
+    }
+
+    fn test_roots() -> RootCertStore {
+        let mut roots = RootCertStore::empty();
+        [
+            include_bytes!("../../../test-ca/ecdsa/ca.der").as_slice(),
+            include_bytes!("../../../test-ca/rsa/ca.der").as_slice(),
+        ]
+        .iter()
+        .for_each(|der| {
+            roots
+                .add(&Certificate(der.to_vec()))
+                .unwrap()
+        });
+        roots
+    }
+
+    #[test]
+    fn test_noauth() {
+        // We should be able to build a verifier that does no client authentication.
+        let builder = WebpkiClientVerifier::builder().allow_unauthenticated();
+        // The builder should be Debug.
+        println!("{:?}", builder);
```

Comment:
Originally I was asserting that the debug string exactly matched expected. I went back to this style because:

a) it felt super fragile to match the debug string output exactly in the tests.
b) when providing real trust anchors (making for better unit tests), the debug output can be quite large. This will also be worse once `webpki::OwnedCertRevocationList` is `Debug`.